Feature/special details

Meesch opened the pull request:
Addresses issue #420 (nice). I did finish this while pretty stuffed with snot, so you might spot some inaccuracies. One thing I was not able to solve was the fact that the help text for has_special_details (in ```study/models.py```) is not picked up by makemessages -a, so it's not translated.... Any idea what's up with that?

Meesch commented:
(also I did a dumb and split this branch off of the hierarchy_fields branch, so it has some commits from that branch that I had to re-commit)

tymees commented:
Hm, something seems to have borked migrations wise....

> Creating test database for alias 'default'...
CommandError: Conflicting migrations detected; multiple leaf nodes in the migration graph: (0024_alter_study_has_traits, 0024_auto_20221018_1653 in studies).
Proceeding without LDAP settings
To fix them run 'python manage.py makemigrations --merge'
Error: Process completed with exit code 1.

Looks like there are two migrations ``0024``

miggol reviewed (changes requested):
Good work on this PR, even though there are still quite a few rough edges. Besides my inline comments I have these general comments as well:

1. The compare page templates need new fields and field logic
2. Now that `traits` are specific to personal _information_, I think the `verbose_name` _"Select the special characteristics of your participants:"_ is a bit confusing. I guess it still makes sense, but maybe we could change that to _"Please specify the kinds of health information you intend to collect:"_ and then rephrase the trait descriptions.

Meesch commented on studies/templates/studies/study_form.html:
```diff
@@ -34,15 +34,16 @@
 
-    depends_on_value('has_traits', 'True', 'traits');
+    depends_on_value('has_special_details', 'True', 'special_details');
+    check_field_required('special_details', 'medical_traits', 'traits', 'studies', 'SpecialDetail');
```
I updated the logic in form_styling.html to make this work

Meesch commented:
Alright, I think I addressed all your comments @miggol. Do please let me know what you think of the dependency logic and whether there are other things I have not yet addressed.

Meesch commented:
Okay, update for the time being: the logic for the field dependencies has been updated, but it glitches if the user changes several fields too quickly after one another. If you wait 4 seconds between toggles of fields, there is no issue, but if you do not, the dependent field will appear and then immediately disappear again. Any insight on why this happens is appreciated, at this point I am ready to rewrite all the dependency logic to solve this.

tymees commented on proposals/templatetags/proposal_filters.py:
```diff
@@ -16,6 +16,15 @@ def needs_details(selected_values, field='needs_details'):
+    result = False
+    for sv in selected_values:
+        if getattr(sv, field):
+            result = True
+            break
```
Better to just return True here, it's a bit cleaner

```python
@register.filter
def medical_traits(selected_values, field='medical_traits'):
    for sv in selected_values:
        if getattr(sv, field):
            return True

    return False

tymees commented:
> Okay, update for the time being: the logic for the field dependencies has been updated, but it glitches if the user changes several fields too quickly after one another. If you wait 4 seconds between toggles of fields, there is no issue, but if you do not, the dependent field will appear and then immediately disappear again. Any insight on why this happens is appreciated, at this point I am ready to rewrite all the dependency logic to solve this.

Yes I can provide some insight;

There are 2 problems working here, both related to code speed.

The first:
The dependency checks using jQuery were unnecessarily slow. This was caused by (re-)searching for the relevant elements in the _entire DOM_. For example:

```js
$('input[type=checkbox]').each( //something );
$('input[type=checkbox]').change();
```
This will do the search for all checkboxes on the page _twice_, which is slow. 
```js
let checkboxes = $('input[type=checkbox]');
checkboxes.each( //something );
checkboxes.change();
```
This will do it once, taking half the time.

As you can guess, this causes some of the problems, as sometimes the checks were taking longer than the speed of the user's clicks.

I've pushed a commit which not only removes the duplicate searches, it also limits the searches where possible to part of the DOM. (Mostly when searching for the checkboxes, which is now limited to the `ul` element that contains them.) This makes the entire process much more speedy. It's still possible to get the problems as before, but it should be way harder.

The second:
This one is similar, but also different, to the first. ``check_field_required`` does a POST request and resolves it asynchronously. This means the JS is unblocked while waiting for the server's response, allowing changes that might invalidate the post request that's being done. Thus, throwing the page into an invalid state. 

I didn't resolve this issue, as I'm unsure on how to tackle that problem right now.

Meesch commented:
> > Okay, update for the time being: the logic for the field dependencies has been updated, but it glitches if the user changes several fields too quickly after one another. If you wait 4 seconds between toggles of fields, there is no issue, but if you do not, the dependent field will appear and then immediately disappear again. Any insight on why this happens is appreciated, at this point I am ready to rewrite all the dependency logic to solve this.
> 
> Yes I can provide some insight;
> 
> There are 2 problems working here, both related to code speed.
> 
> The first: The dependency checks using jQuery were unnecessarily slow. This was caused by (re-)searching for the relevant elements in the _entire DOM_. For example:
> 
> ```js
> $('input[type=checkbox]').each( //something );
> $('input[type=checkbox]').change();
> ```
> 
> This will do the search for all checkboxes on the page _twice_, which is slow.
> 
> ```js
> let checkboxes = $('input[type=checkbox]');
> checkboxes.each( //something );
> checkboxes.change();
> ```
> 
> This will do it once, taking half the time.
> 
> As you can guess, this causes some of the problems, as sometimes the checks were taking longer than the speed of the user's clicks.
> 
> I've pushed a commit which not only removes the duplicate searches, it also limits the searches where possible to part of the DOM. (Mostly when searching for the checkboxes, which is now limited to the `ul` element that contains them.) This makes the entire process much more speedy. It's still possible to get the problems as before, but it should be way harder.
> 
> The second: This one is similar, but also different, to the first. `check_field_required` does a POST request and resolves it asynchronously. This means the JS is unblocked while waiting for the server's response, allowing changes that might invalidate the post request that's being done. Thus, throwing the page into an invalid state.
> 
> I didn't resolve this issue, as I'm unsure on how to tackle that problem right now.

Nice! This is indeed faster and better, thank you for explaining. As for the second issue: I must confess I do not fully understand, but I am happy to hear your in-person explanation of the problem!

miggol reviewed (approved):
I found some missing translations and added them right away. This is now ready for merge as far as I'm concerned. Thanks @tymees for chipping in on the JS side of things.

tymees reviewed (approved):
I've changed the request handling a bit to use a map, so we can keep track of requests for multiple fields at the same time.

Otherwise, a user might still click fast enough to mess up our code. 

I agree that it's mergeable now :)
